token-metadata/program/src/instruction/metadata.rs

```diff
@@ -100,20 +100,278 @@ pub enum UpdateArgs {
         /// Required authorization data to validate the request.
         authorization_data: Option<AuthorizationData>,
     },
+    UpdateAuthorityV2 {
+        /// The new update authority.
+        new_update_authority: Option<Pubkey>,
+        /// The metadata details.
+        data: Option<Data>,
+        /// Indicates whether the primary sale has happened or not (once set to `true`, it cannot be
+        /// changed back).
+        primary_sale_happened: Option<bool>,
+        // Indicates Whether the data struct is mutable or not (once set to `true`, it cannot be
+        /// changed back).
+        is_mutable: Option<bool>,
+        /// Collection information.
+        collection: CollectionToggle,
+        /// Additional details of the collection.
+        collection_details: CollectionDetailsToggle,
+        /// Uses information.
+        uses: UsesToggle,
+        // Programmable rule set configuration (only applicable to `Programmable` asset types).
+        rule_set: RuleSetToggle,
+        /// Token standard.
+        token_standard: Option<TokenStandard>,
+        /// Required authorization data to validate the request.
+        authorization_data: Option<AuthorizationData>,
+    },
+    AuthorityItemDelegateV2 {
```

Comment:
Why do we need the `Delegate` word in the name for the above ones? Are they not just args?

Reply:
Having another thought about this after looking into the whole PR. What if the `V2` version had an inner struct that represent the fields to be updated instead of multiple args versions? E.g.,:
```
pub enum UpdateArgs {
    V1 { ... },
    V2 {
        fields: UpdateFields,
        authorization_data: Option<AuthorizationData>,
    }
}

pub enum UpdateFields {
    Collection {
        /// Collection information.
        collection: CollectionToggle,
    },
    Data {
        /// The metadata details.
        data: Option<Data>,
    }
    ...
}
```

Reply:
+1 one on removing the `Delegate` name. This will greatly affect the generated API for the end-user.

Right now a `AuthorityItemDelegateV2` variant within the `Update` handler will generate a method named: `updateAuthorityItemDelegateV2` which is super confusing.

Regarding nested enums: That's possible on Kinobi but the generated methods will be called "handler" + "variantLevel1" + "variantLevel2". So in the provided example, we'll have `updateV2Collection` and `updateV2Data` for instance.

Reply:
Discussed with @febo, @lorisleiva, @blockiosaurus, and @samuelvanderwaal offline and we decided to keep the overall `UpdateArgs` struct as the same one-level-deep, struct-like enum variants, use "V2" in the name rather than "V1", and also change the names to include the word "As" to indicate that the caller is doing the metadata update "as" the specified role:

```Rust
pub enum UpdateArgs {
    ...
    AsUpdateAuthorityV2 {
        /// The new update authority.
        new_update_authority: Option<Pubkey>,
        ...
    },
    AsAuthorityItemDelegateV2 {
        /// The new update authority.
        new_update_authority: Option<Pubkey>,
        ...
    },
    AsCollectionDelegateV2 {
        /// Collection information.
        collection: CollectionToggle,
        ...
    },
    AsDataDelegateV2 {
        /// The metadata details.
        data: Option<Data>,
       ...
    },
```

We chose this because:
* The one-level-deep struct-like variants is consistent with other instruction args types, such as `DelegateArgs`.
* It will result in better generated typescript names.
* It will make it clear to the caller how its working.  For example, they are updating "as the update authority", **not** that they are updating the update authority value itself.
* In the future we didn't want to have multiple versions in the same overall args, something like `V3 { fields: UpdateFieldsV2, ...}`.

We chose to keep it at V2 instead of  V1 and we also chose to not abbreviate "Programmable" in some of the args.